### PR TITLE
spirv-val: Validate outter descriptor array index

### PR DESCRIPTION
**Pull request by spencer-lunarg**
I found out that `spirv-val` wasn't validating access chain OOB for the outer array of a descriptor array

```glsl
// buffer array
layout(set=0, binding=0) buffer foo { int x; } var[3];
var[3].x = 3;

// image array
layout(set = 0, binding = 0) uniform sampler sampler_array[2];
sampler2D(texture_image, sampler_array[2])
```

### CODE REVIEW

File: source/val/validate_memory.cpp [hunk 1]
```diff
@@ -1276,14 +1276,13 @@ spv_result_t ValidateCopyMemory(ValidationState_t& _, const Instruction* inst) {
 
 spv_result_t ValidateAccessChain(ValidationState_t& _,
                                  const Instruction* inst) {
-  std::string instr_name =
```
Comment (spencer-lunarg) [inline source/val/validate_memory.cpp:1279]: moved string logic to error message time as assume most `OpAccessChain`  are valid and don't need to spend time building this string

File: source/val/validate_memory.cpp [hunk 2]
```diff
@@ -1356,27 +1356,46 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
...
         type_pointee = _.FindDef(type_pointee->word(2));
         break;
       }
+      case spv::Op::OpTypeArray: {
+        // If using a descriptor array, make sure the outer array is not OOB
+        if (spv::Op::OpConstant == cur_word_instr->opcode()) {
+          const uint32_t cur_index = cur_word_instr->word(3);
+          auto array_length_id = _.FindDef(type_pointee->word(3));
+          if (spv::Op::OpConstant == array_length_id->opcode()) {
+            const uint32_t array_length = array_length_id->word(3);
```
Comment (spencer-lunarg) [inline source/val/validate_memory.cpp:1379]: while the `OpAccessChain` indexes are required to be 32-bit int scalar but can be negative

I realize from the failing `spirv-opt` tests is the `OpTypeArray` length operand can be a 64-bit length

File: source/val/validate_memory.cpp [hunk 3]
```diff
@@ -1356,27 +1356,46 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
...
       case spv::Op::OpTypeStruct: {
         // In case of structures, there is an additional constraint on the
         // index: the index must be an OpConstant.
         if (spv::Op::OpConstant != cur_word_instr->opcode()) {
           return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
-                 << "The <id> passed to " << instr_name
+                 << "The <id> passed to Op" << spvOpcodeString(opcode)
                  << " to index into a "
                     "structure must be an OpConstant.";
         }
```
Comment (spencer-lunarg) [inline source/val/validate_memory.cpp:1401]: @alan-baker working more into this, I found the line below makes it seem like it is legal to have a `-1` access chain index?

So this means if I have a 

`layout(set=0, binding=0) buffer foo { int x[3]; } var;`

These are valid?

```swift
%int_n1 = OpConstant %int -1
%int_5 = OpConstant %int 5

%ac = OpAccessChain %ptr %var %int_0 %int_n1
// or
%ac = OpAccessChain %sb_ptr %var %int_0 %int_5
```

**Comment by s-perron** [comment]
Are we sure we want this in the validator? Having an OpAccessChain that is oob is not invalid, it is [undefined behaviour](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpAccessChain:~:text=%2D%20if%20indexing%20into%20a%20vector%2C%20array%2C%20or%20matrix%2C%20with%20the%20result%20type%20being%20a%20logical%20pointer%20type%2C%20causes%20undefined%20behavior%20if%20not%20in%20bounds.), if it is executed.

This is important because someone might have code that is guarded by a conditional statement that will only execute it when it is allowed. This seems like a better fit for the linter.

**Comment by spencer-lunarg** [comment]
@s-perron I guess the issue was someone had the shader in https://github.com/KhronosGroup/SPIRV-Tools/issues/5468

and it seemed like something that could be caught statically so it was not be validated in the validation layers SPIR-V runtime

We are currently checking if some of the `OpAccessChain` is being accessed correctly in `spirv-val`, so it was strange it was not for all aspects

**Comment by s-perron** [comment]
> We are currently checking if some of the OpAccessChain is being accessed correctly in spirv-val

That surprises me because I thought we had agreed it was not a validation error.

**Comment by alan-baker** [comment]
> > We are currently checking if some of the OpAccessChain is being accessed correctly in spirv-val
> 
> That surprises me because I thought we had agreed it was not a validation error.

When I look at the code we currently on check OpTypeStruct indexing. That is specifically called out as *must* (validity) in SPIR-V.

As I mentioned in the other issue, I would be ok with Vulkan specific checks that error on guaranteed undefined behaviour. I agree with Steven that that means you might have to limit where those checks are applied (e.g. always executed).

**Comment by spencer-lunarg** [comment]
> When I look at the code we currently on check OpTypeStruct indexing. That is specifically called out as must (validity) in SPIR-V.

Ok, good to know!

>  I agree with Steven that that means you might have to limit where those checks are applied (e.g. always executed).

Ok, let me take another stab at this, while we do plan to have GPU Validation on OOB, there is a large amount of trivial cases that could be caught, happy to guard (and comment in the code) the restriction of the checking

**Comment by spencer-lunarg** [comment]
After spending more time on this, decided that 

1. GLSL/HLSL/Slang don't even allow you do do this, so it would be someone code gen going bad
2. I prototyped this in VVL and it was quite a lot of code/overhead for something that is not likely to hit
3. GPU-AV will catch this anyway just by nature of it being OOB

closing and just gonna keep this as a runtime check

**Pull request closed without merging**